custom_components/frigate/api.py

```diff
@@ -216,27 +228,101 @@ async def async_get_recordings(
         )
         return cast(dict[str, Any], result) if decode_json else result
 
+    async def _get_token(self) -> None:
+        """
+        Obtain a new JWT token using the provided username and password.
+        Sends a POST request to the login endpoint and extracts the token
+        and expiration date from the response headers.
+        """
+        response = await self.api_wrapper(
+            method="post",
+            url=str(URL(self._host) / "api/login"),
+            data={"user": self._username, "password": self._password},
+            decode_json=False,
+            login_request=True,
+        )
+
+        set_cookie_header = response.headers.get("Set-Cookie", "")
+        if not set_cookie_header:
+            raise KeyError("Missing Set-Cookie header in response")
+
+        token_found = False
+        for cookie_prop in set_cookie_header.split(";"):
+            cookie_prop = cookie_prop.strip()
+            if cookie_prop.startswith("frigate_token="):
+                jwt_token = cookie_prop.split("=", 1)[1]
+                self._token_data["token"] = jwt_token
+                token_found = True
+
+                try:
+                    decoded_token = jwt_wrapper.unverified_hs256_token_decode(jwt_token)
+                except Exception as e:
+                    raise ValueError(f"Failed to decode JWT token: {e}")
+                exp_timestamp = decoded_token.get("exp")
+                if not exp_timestamp:
+                    raise KeyError("JWT is missing 'exp' claim")
+                self._token_data["expires"] = datetime.datetime.fromtimestamp(
+                    exp_timestamp, datetime.UTC
+                )
+
+        if not token_found:
+            raise KeyError("Missing 'frigate_token' in Set-Cookie header")
+
+    async def _refresh_token_if_needed(self) -> None:
+        """
+        Refresh the JWT token if it is expired or about to expire.
+        """
+        if "expires" not in self._token_data:
+            await self._get_token()
+            return
+
+        current_time = datetime.datetime.now(datetime.UTC)
+        if current_time >= self._token_data["expires"]:  # Compare UTC-aware datetimes
+            await self._get_token()
+
+    async def _get_headers(self) -> dict[str, str]:
```

Comment:
Perhaps `_get_auth_headers` would be more descriptive here, as I don't think there are any other headers involved.